src/test/run-make/sanitizer-staticlib-link/Makefile

```diff
@@ -0,0 +1,25 @@
+-include ../tools.mk
+
+# This test builds a staticlib, then an executable that links to it.
+# The staticlib and executable both  are compiled with address sanitizer, 
+# and we assert that a fault in the staticlib is correctly detected.
+
+# NOTE the address sanitizer only supports x86_64 linux and macOS
+
+ifeq ($(TARGET),x86_64-apple-darwin)
+ASAN_SUPPORT=$(SANITIZER_SUPPORT)
+EXTRA_RUSTFLAG=-C rpath
+else
+ifeq ($(TARGET),x86_64-unknown-linux-gnu)
+ASAN_SUPPORT=$(SANITIZER_SUPPORT)
+EXTRA_RUSTFLAG=
+endif
+endif
+
+all:
+ifeq ($(ASAN_SUPPORT),1)
+	$(RUSTC) -g -Z sanitizer=address --crate-type staticlib --target $(TARGET) library.rs
+	$(CC) program.c $(call STATICLIB,library) $(call OUT_EXE,program) -lasan $(EXTRACFLAGS) $(EXTRACXXFLAGS)
```

Comment:
I don't think that -lrustc_asan will work because there is no .so of rustc_asan compiled on my system. I think perhaps the issue is the depkind is implicit not explicit when we link the runtime. From the original addition: https://github.com/rust-lang/rust/pull/38699/files#diff-974185a1d0122c6d970a1e4faaedbe5fR834

So I'm trying the build again but with the explicit dep, so hopefully that should force the addition for exec and static (and leave it alone for the others). Saying this, I'm not an expert on the rustc internals, so I could be barking up the wrong tree :)